### PR TITLE
Fixes #23831: Migrate change request API in change-validation to zio-json

### PR DESCRIPTION
**Pull request by clarktsiory**
https://issues.rudder.io/issues/23831

This PR migrates only the zio-json parts of the change request APIs. It consists in : 
* adding tests to the API and the database which is left untouched
* removing usage of `RestExtractorService` from Rudder which is the only part using lift-json indirectly

A whole new ADT tree needed to be created to be in par with the domain model of `ChangeRequest`, and we heavily use _chimney_ to do the mapping, see the [ChangeRequestJson.scala](https://github.com/Normation/rudder-plugins/blob/eebed5083c2ba840ef6919314938d7e5b8178c4b/change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJson.scala) file. This time, we use the `PartialTransformer` from _chimney_ because serialization could result in errors anywhere in the transformations. 

The behavior is slightly changed when encountering errors when attempting to serialize a change request : 
* previously we had serialized an "error string" in place of a serialized `Directive` object to encode the error
* now we fail-fast as early as we can before even serializing (e.g. when fetching the context to serialize the `Directive`) and we accumulate any other serialization errors in the error channel, which is provided by _chimney_ and which we adapt to our app's.

There may still be some works to do to for the migration : 
* the lift `Box` could be rewritten to `IOResult` effect type in most places, even up to Rudder interfaces
* the errors messages could be tested and improved further : _chimney_ tracks the path or errors, so when doing the one partial transformation of a `ChangeRequest` we could have very details errors in nested fields.

### CODE REVIEW

**Comment by clarktsiory** [comment]
Commit modified

**Comment by clarktsiory** [comment]
Commit modified

**Comment by clarktsiory** [comment]
PR updated with a new commit

**Comment by clarktsiory** [comment]
PR updated with a new commit

**Comment by clarktsiory** [comment]
PR updated with a new commit

**Comment by clarktsiory** [comment]
PR updated with a new commit

**Comment by clarktsiory** [comment]
PR updated with a new commit

File: change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+                      TechniqueName("packageManagement"),
+                      mockDirectives.directives.pkgDirective.copy(
+                        name = "pkg_directive_001",
+                        shortDescription = "testing directive change",
+                        longDescription = "testing directive change",
+                        priority = 1,
+                        isSystem = true,
+                        parameters = mockDirectives.directives.pkgDirective.parameters ++ Map(("PACKAGE_LIST", Seq("curl")))
+                        // ignored changes :
+                        // policyMode = Some(PolicyMode.Audit),
```
Comment (fanf) [inline change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala:299]: policy mode should not be ignored. Why is it so ? 
(even tag, I'm not sure) (they should, they are in eventlog: https://issues.rudder.io/issues/15019)

File: change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+                      name = "rule_002",
+                      shortDescription = "testing rule change",
+                      longDescription = "testing rule change",
+                      directiveIds = Set(DirectiveId(DirectiveUid("directive1"))),
+                      targets = Set(AllTargetExceptPolicyServers),
+                      isEnabledStatus = false,
+                      isSystem = false
+                      // ignored changes :
+                      // categoryId = RuleCategoryId("rule_category_001"),
+                      // policyMode = Some(PolicyMode.Audit),
```
Comment (fanf) [inline change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala:486]: same: I don't see why these change are not tracked. PolicyMode at least has clear impact on prod. 
Can you look to that and potentially open a ticket in the correct branch if needed ?

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import com.normation.rudder.AuthorizationType
+import com.normation.rudder.Rights
+import com.normation.rudder.web.services.RudderUserDetail
+import net.liftweb.http.SessionVar
+import org.springframework.security.core.context.SecurityContextHolder
+
+/**
+  * Redefine the Rights methods as in Rudder, constructor allows to provide RudderUserDetail
+  */
+class CurrentUser protected (get: => Option[RudderUserDetail]) extends SessionVar[Option[RudderUserDetail]](get) {
```
Comment (fanf) [inline change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala:49]: You absolutly don't want to define these kind of security/authorization thing at several place. If it exists in Rudder, you must not ever define it elsewhere, it's a sure path toward inconsistancy and security hole. 

If you are missing something with Current user: 
- either you change it in Rudder, and ensure that everything still work as expected, 
- or you use adapt for test (if it's only a proble of test). 

Plus: Rudder `CurrentUser` has a `set` method, so you can `CurrentUser.set(Some(userDetails))` if needed

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJson.scala [hunk 4]
```diff
@@ -0,0 +1,961 @@
...
+    .withFieldComputed(_.id, _.id.serialize)
+    .withFieldRenamed(_.name, _.displayName)
+    .withFieldComputed(_.categoryId, _.categoryId.value)
+    .withFieldComputed(_.directives, _.directiveIds.map(_.serialize).toList.sorted)
+    .withFieldComputed(_.targets, _.targets.toList.sortBy(_.target).map(t => JRRuleTarget(t)))
+    .withFieldRenamed(_.isEnabledStatus, _.enabled)
+    .withFieldComputed(_.tags, x => JRTags.fromTags(x.tags))
+    .withFieldConst(_.policyMode, None)      // not needed
+    .withFieldConst(_.status, None)          // not needed
+    .withFieldConst(_.changeRequestId, None) // not needed
```
Comment (fanf) [inline change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJson.scala:490]: I think they should. If they are not in 7.3/8.0, it's a bug - can you check and do what is needed accordingly ?

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJson.scala [hunk 5]
```diff
@@ -0,0 +1,961 @@
...
+import io.scalaland.chimney.syntax._
+import net.liftweb.common.Empty
+import net.liftweb.common.Failure
+import net.liftweb.common.Full
+import scala.util.Try
+import zio.Chunk
+import zio.NonEmptyChunk
+import zio.json._
+import zio.json.internal.Write
+
```
Comment (fanf) [inline change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJson.scala:112]: That seems to have been a really easy piece of code to craft :D

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala [hunk 6]
```diff
@@ -113,11 +113,29 @@ class EmailNotificationService {
...
+}
+
+object NotificationService {
+  def apply(
+      emailService:   EmailNotificationService,
+      linkUtil:       LinkUtil,
+      configMailPath: String
+  ): NotificationService = {
+    new NotificationServiceImpl(
+      emailService,
```
Comment (fanf) [inline change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala:127]: I think I'm against that idea - but I'm not sure to understand the goal. Is it a kind of default implementation ? 
It looks like it adds a lot of burden in maintenance (each change in constructor have to been done at one more place) and I fail to see the earning. Can you explain ?

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/WorkflowJdbcRepository.scala [hunk 7]
```diff
@@ -63,36 +63,60 @@ trait WoWorkflowRepository {
...
+  def createWorkflowSQL(crId: ChangeRequestId, state: WorkflowNodeId): Update0 = {
+    sql"insert into workflow (id, state) values ($crId, $state)".update
+  }
+
+  def updateStateSQL(crId: ChangeRequestId, from: WorkflowNodeId, state: WorkflowNodeId): Update0 = {
+    sql"update workflow set state = $state where id = $crId".update
+  }
+}
+trait WorkflowJdbcRepositorySQL extends RoWorkflowJdbcRepositorySQL with WoWorkflowJdbcRepositorySQL {}
+object WorkflowJdbcRepositorySQL extends WorkflowJdbcRepositorySQL
```
Comment (fanf) [inline change-validation/src/main/scala/com/normation/plugins/changevalidation/WorkflowJdbcRepository.scala:90]: Can you move `WorkflowJdbcRepositorySQL` into test since it's not use here?
Plus, I think the trait `WorkflowJdbcRepositorySQL` isn't use, you could just have in the test file `object WorkflowJdbcRepositorySQL extends RoWorkflowJdbcRepositorySQL with WoWorkflowJdbcRepositorySQL {}` no ?

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/api/ChangeRequestApi.scala [hunk 8]
```diff
@@ -93,87 +96,94 @@ object ChangeRequestApi       extends ApiModuleProvider[ChangeRequestApi] {
...
     readChangeRequest:    RoChangeRequestRepository,
     writeChangeRequest:   WoChangeRequestRepository,
     readWorkflow:         RoWorkflowRepository,
-    writeWorkflow:        WoWorkflowRepository,
-    readTechnique:        TechniqueRepository,
     workflowLevelService: WorkflowLevelService,
     commitRepository:     CommitAndDeployChangeRequestService,
-    restDataSerializer:   RestDataSerializer
+    userPropertyService:  UserPropertyService,
+    currentUser:          CurrentUser
```
Comment (fanf) [inline change-validation/src/main/scala/com/normation/plugins/changevalidation/api/ChangeRequestApi.scala:166]: you should not pass `CurrentUser` in a constructor. `CurrentUser` is an entity (a strange one, with a very strange persistance and retrieval model) but it's still an entity. 
Only parameters or other services are passed to service constructor. If you want to pass something to get current user (which is ok, it's not great to directly refer a magic entity in code), you should pass `com.normation.rudder.UserService`

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/api/ChangeRequestApi.scala [hunk 9]
```diff
@@ -277,45 +264,14 @@ class ChangeRequestApiImpl(
...
-      } catch {
-        case e: Exception =>
-          toJsonError(None, JString(s"'${sid}' is not a valid change request id (need to be an integer)"))(
-            "changeRequestDetails",
-            restExtractor.extractPrettify(req.params)
-          )
-      }
+      withChangeRequestContext(sid, params, schema, "find")((changeRequest, status, techniqueByDirective) =>
+        serialize(changeRequest, status)(techniqueByDirective).toIO
+      ).toLiftResponseOne(params, schema, Some(sid))
```
Comment (fanf) [inline change-validation/src/main/scala/com/normation/plugins/changevalidation/api/ChangeRequestApi.scala:269]: 👌

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJson.scala (hunk 4 above)
Comment (clarktsiory) [reply change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJson.scala:490]: Here specifically these fields are used when there is a "new" Rule change request or a "deletion" of a Rule change request (`JRRule` is only used for serializing the resulting CR in both cases). 

The previous lift-json code replaced by this one does not in fact serialize these fields : 
```
  def serializeRule(rule: Rule, crId: Option[ChangeRequestId]): JValue = {
    (("changeRequestId"   -> crId.map(_.value.toString))
    ~ ("id"               -> rule.id.uid.value)
    ~ ("revision"         -> rule.id.rev.toOptionString)
    ~ ("displayName"      -> rule.name)
    ~ ("categoryId"       -> rule.categoryId.value)
    ~ ("shortDescription" -> rule.shortDescription)
    ~ ("longDescription"  -> rule.longDescription)
    ~ ("directives"       -> rule.directiveIds.toList.map(_.serialize).sorted)
    ~ ("targets"          -> rule.targets.map(_.toJson))
    ~ ("enabled"          -> rule.isEnabledStatus)
    ~ ("system"           -> rule.isSystem)
    ~ ("tags"             -> serializeTags(rule.tags)))
  }
```

A way to avoid this confusion is the create another case class like `JRRule` but with only those serialized fields, but I chose to reuse `JRRule`, it has all we need (even more)

File: change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala (hunk 1 above)
Comment (clarktsiory) [reply change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala:299]: Both `policyMode` and `tags` changes would be taken into account : https://github.com/Normation/rudder/blob/master/webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/policies/DirectiveDiff.scala#L86, but in the API response (yaml tests), the diff for both of them were never displayed : https://github.com/Normation/rudder/blob/master/webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/apidata/RestDataSerializer.scala#L535

It could be added without risk, but I didn't because it would mean I'd also have to add these new fields to the 'json datatype' and also change the json tests... Is it okay to keep those lines as comments or should I make the change ?

File: change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala (hunk 2 above)
Comment (clarktsiory) [reply change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala:486]: Same answer too, the diff is actually taken into account, here the previous version did not 'serialize' the change for these fields, so I kept that behavior, but it could be changed without risk

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala (hunk 6 above)
Comment (clarktsiory) [reply change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala:127]: As I wrote a new interface `NotificationService` to be able to use a mock version of it in tests, I renamed the class to `NotificationServiceImpl`, so previously `new NoficationService(args...)` would be now just `NotificationService(args...)`. It was either `NotificationService(args...)` everywhere, or `new NotificationServiceImpl(args...)`. I prefered the first version as it was returning an actual `NotificationService`, the trait and not the implementation.  

It also can be seen as a constructor that hides the implementation, but now the impl remains public (because the actual implementation is tested, including methods not in the interface...) so it kinda defeats its own purpose because of that :sweat_smile:

**Comment by clarktsiory** [comment]
PR updated with a new commit

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala (hunk 3 above)
Comment (clarktsiory) [reply change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala:49]: :+1: I didn't think the `set` method existed, it simplifies a lot of things ! (morale : sometimes mutability does, and we need it sometimes) :smile:

Now I only have to wait for this PR to make the `set` method available : https://github.com/Normation/rudder/pull/5267

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/WorkflowJdbcRepository.scala (hunk 7 above)
Comment (clarktsiory) [reply change-validation/src/main/scala/com/normation/plugins/changevalidation/WorkflowJdbcRepository.scala]: The solution was as you suggested to remove the unnecessary trait, the `object  WorkflowJdbcRepositorySQL` is used in the file here (to implement the repos) as well as in query type-checking tests, so there is only a single object that contains all the SQL now : https://github.com/Normation/rudder-plugins/pull/618/commits/3fbb60377fcd6ac1c278b9f60319f44797af6e1f

**Comment by clarktsiory** [comment]
PR rebased

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala (hunk 3 above)
Comment (clarktsiory) [reply change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala:49]: Now that that one PR is merged I could not do the test without this other one :smile:  : https://github.com/Normation/rudder/pull/5289 (`set`-ting the user is not possible at top-level, we need to do it in each yaml test so it needs a change in the test trait in Rudder)

**Comment by clarktsiory** [comment]
PR updated with a new commit

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala (hunk 3 above)
Comment (clarktsiory) [reply change-validation/src/main/scala/com/normation/plugins/changevalidation/CurrentUser.scala]: A simpler and better solution : rely on the `UserService`, which allows indeed to have a user with the correct rights in tests : https://github.com/Normation/rudder-plugins/pull/618/commits/9d45951ef7ded84ca44692a0fd84f2c7aa87f8cd

With that last commit I should have addressed all issues for this PR...

**Comment by clarktsiory** [comment]
PR updated with a new commit

File: change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala (hunk 1 above)
Comment (clarktsiory) [reply change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala:299]: I think I would be better off in another PR, because I tried to implement it and the change needs a consequent adaptation of the current code (adapt the `{"from":...,"to":...}` json serialisation because it does not display 'null', and add corresponding lines of code/tests)

File: change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala (hunk 2 above)
Comment (clarktsiory) [reply change-validation/src/test/scala/com/normation/plugins/changevalidation/api/ChangeRequestApiTest.scala:486]: Cf. the comment above (I think I would be better off in another PR)

**Comment by clarktsiory** [comment]
PR updated with a new commit

File: change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala (hunk 6 above)
Comment (clarktsiory) [reply change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala]: Resolved in https://github.com/Normation/rudder-plugins/pull/618/commits/e3e6696e56805494aa10d21a8202cf95284744bf !

**Comment by clarktsiory** [comment]
Still rebasing (lasts rebase hopefully :crossed_fingers:)

**Review by fanf — APPROVED** [review]
LGTM.
Not exactly a small pr, I hope the remaining lift rest endpoints will be easier to port
